### PR TITLE
tox tests on github actions

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This pr would move tests to github actions backed by [tox](https://tox.readthedocs.io/en/latest/) environments.

Tox is an awesome tool that lets you declare environments, dependencies, and commands for testing.  It lets you define a number of environments, or environmental factors, and then mix and match them when calling on the command line.  For instance, `tox -e py37-pyside` will build you a new environment with python 3.7 and pyside2 and run all the tests (i.e. you don't need to have create one yourself with conda).  It also makes sure that you're testing in a "clean" environment, as would be achieved in a fresh install, without pollution from your dev environment.  Even better, run `tox -p` and _all_ of the platform-appropraite environments in your [tox.ini envlist](https://tox.readthedocs.io/en/latest/config.html#conf-envlist) will be run in parallel. This makes it super easy to run environmentally parametrized tests locally.  (Note, if you want `tox` to build your environments with conda instead of virtualenv, install [tox-conda](https://github.com/tox-dev/tox-conda)).

It plays really well with github actions via [tox-gh-actions](https://github.com/ymyzk/tox-gh-actions), where various environmental variables can be used to select the appropriate [tox factors](https://tox.readthedocs.io/en/latest/config.html#factors-and-factor-conditional-settings), limiting the scope of the tests based on the test environment.

tests are also now working on py3.9 on all platforms, windows pyqt5 (which wasn't before), and big sur with py3.9.0

I've broken up the tests into three typical triggers (just a proposal, very much open for feedback):
1. A quick (~10min) but "thorough-enough" test that gets run on pull requests: [`test_pull_requests.yml`](https://github.com/tlambert03/napari/blob/gh-actions/.github/workflows/test_pull_requests.yml).  This tests:
   - code quality and manifest.in
   - py37,38,39 on ubutuntu with both pyside2 and pyqt5
   - windows py3.8 with both pyside2 and pyqt5
   - macos py.38 with only pyqt5 (mac is the slowest, so this is just to save about 5 minutes)
   - ubuntu python 3.7 with MIN_REQ
2. A "comprehensive" test that takes somewhere between 20-40 min (mostly because Big Sur takes forever to start... we could exclude that) that runs on any push to `master`: [`test_comprehensive.yml`](https://github.com/tlambert03/napari/blob/gh-actions/.github/workflows/test_comprehensive.yml)
   - py37,38,39 on mac,win,linux with both pyqt5 and pyside2
   - python 3.9.0 on big sur with pyqt5
   - ubuntu python 3.7 with MIN_REQ
   - ubuntu python 3.8 and pyqt5 with `--async_only`
   - ubuntu python 3.8 and pyqt5 with `--perfmon-only`
3. A `pip --pre` test that runs on a cron job every 12 hours: [`test_prereleases.yml`](https://github.com/tlambert03/napari/blob/gh-actions/.github/workflows/test_prereleases.yml).  This tests mac, windows, and linux on python3.8 with both pyside2 and pyqt5 and opens an issue if anything fails.  [Here is an example](https://github.com/tlambert03/napari/issues/26) of what a failing `--pre` test issue looks like.

~**I've temporarily enabled all three on pull requests, so they should run when I submit this, but you can also see my last runs here: https://github.com/tlambert03/napari/actions**~ (I put them back to their triggers)

### other notes:
- ~I bumped numpy to >= 1.16.5 to pass the `MIN_REQS` test... but that's just for pandas I think, so we can revert that if we want to figure out a pandas-specific strategy~ (taken care of in #2050)
- ~I'm including a test fix from #1856 in the magicgui tests~ (merged in #1856)
  
## Type of change

- [x] Changing test runner

### CODE REVIEW

**Comment by tlambert03** [comment]
one more note: you'll notice that the three different workflows are basically only different in their test matrices, but github [doesn't currently seem to offer](https://github.community/t/how-to-conditionally-include-exclude-items-in-matrix-eg-based-on-branch/16853/6) conditional matrices.  However, we could potentially [this strategy](https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#example-6) to construct a matrix as a build step based on the event trigger... and then merge all three files.

**Comment by sofroniewn** [comment]
This is great @tlambert03! I'll spend some time digesting, but I'm supportive. One quick note is it looks like the "Black Formatting" test is hanging - see 
image

**Comment by tlambert03** [comment]
> This is great @tlambert03! I'll spend some time digesting, but I'm supportive. One quick note is it looks like the "Black Formatting" test is hanging - see
> image

yeah, we'll need to go into Settings > Branches > `master` > ... Status checks ... to change that before merging.
(It's not hanging, it's waiting on a test that no longer exists)

**Comment by jni** [comment]
Holy moly LOL. (Looking at the number of comprehensive tests. 😅) @tlambert03 how do notifications work on failing tests on master and pre? I think this is technically the correct way to split tests, but I'm worried that no one will click on that gh actions tab to check on this stuff. PRs are nice because there's always people noticing failing checks.

Anyway, other than that I love this change. =)

**Comment by tlambert03** [comment]
> how do notifications work on failing tests on master and pre? I think this is technically the correct way to split tests, but I'm worried that no one will click on that gh actions tab to check on this stuff.

The `--pre` test will automatically open an issue if one of the tests fail. ([Here is an example](https://github.com/tlambert03/napari/issues/26) of what a failing `--pre` test issue looks like).

The comprehensive tests are usually a bit more visible (since they will have a big X on the repo home page). but there's no reason I can't/shouldn't add the same issue-opening mechanism to that too

**Comment by tlambert03** [comment]
ok, I put the comprehensive and `--pre` tests back to their "normal" settings of running on push-to-master and cron job respectively.  both will not open an issue when a test fails.  You can also trigger each of them manually in the actions tab (this is because of the [`workflow_dispatch` event](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/) that I added ... which we might want to do for the bundle since we sometimes need to trigger that manually?

I also added python 3.9 to setup config, since this pretty much confirms we can "officially" support it now.

Lastly, I tried to add a lot of comments to each section explaining how it all works for whoever needs to change it in the future.

the last thing we'd need here is to switch the "Black Formatting" requirement in Settings > Branches > master > ... Status checks ... to `Code Quality (black)`

File: .github/workflows/test_pull_requests.yml
Comment (sofroniewn) [inline .github/workflows/test_pull_requests.yml]:
```suggestion
name: Quick
```

Comment (sofroniewn) [reply .github/workflows/test_pull_requests.yml]: Its just so many letters on the dialog otherwise :-) 
image

File: .github/workflows/test_comprehensive.yml
Comment (sofroniewn) [inline .github/workflows/test_comprehensive.yml]:
```suggestion
name: Comprehensive
```

File: .github/workflows/test_prereleases.yml
Comment (sofroniewn) [inline .github/workflows/test_prereleases.yml]:
```suggestion
name: Pre-releases
```

File: .github/workflows/test_comprehensive.yml
Comment (sofroniewn) [inline .github/workflows/test_comprehensive.yml:77]: Can we add these back to the `Quick` test too? We've still got an async PR in progress and it would be nice to know as we refactor that we aren't breaking things. We can drop from `Quick` in the future if we feel excessive.

**Review by sofroniewn — APPROVED** [review]
Ok, this looks good to me. Great work @tlambert03. I'm approving, have made a couple fairly minor name requests and adding `asycn` / `perfmon` back to PR tests (at least for now).

I'd like @jni to approve too before we merge as this is a big change. But I'm good for it to go in anytime now!

File: .github/workflows/test_comprehensive.yml
Comment (tlambert03) [reply .github/workflows/test_comprehensive.yml:77]: Sure can do.  Can someone help me understand what's going on with the perfmon test?  I don't get why we have a dedicated test in the matrix just to [execute a single test function and skip all of the others](https://cirrus-ci.com/task/6286878838095872?command=test_pyqt5#L1655)... is just to avoid mocking USE_PERFMON for that test?

Comment (sofroniewn) [reply .github/workflows/test_comprehensive.yml:77]: Yeah, I think the we should drop `pytest.mark.perfmon`. I think at the time it was thought we might end up with more, but it doesn't make sense to have all the extra machinery for one test. I'll look into making a new PR to drop tonight. I think there are more for `ASYNC`

File: .github/workflows/test_pull_requests.yml
Comment (tlambert03) [reply .github/workflows/test_pull_requests.yml]: agreed. I'll shorten them all, but I'd like to leave in "Test" (since we also have docs, bundle, etc...) so how about `PR Test`, `Full Test`, `--pre Test` ... and I can shorten `Code Quality` too

Comment (sofroniewn) [reply .github/workflows/test_pull_requests.yml]: those are fine names, i was just trying to trim down a little, don't have strong feelings

**Comment by sofroniewn** [comment]
Thanks for moving perfmon/ async over. I'll try and get rid of perfmon as I said above. Didn't get to it tonight, but we can also do that after this merges too. If @jni gives this an approve then I think we're good to merge !!!!

**Comment by sofroniewn** [comment]
Now that #2055 has merged can you add a `no-Qt` / `Qt isolation` test @tlambert03. Probably something that should be part of `Quick` but only linux.

**Comment by tlambert03** [comment]
headless tests added: https://github.com/napari/napari/runs/1631288034?check_suite_focus=true#step:7:19

**Review by jni — APPROVED** [review]
Very very awesome @tlambert03! :tada: The issue auto-opening is mind-blowing. So happy about it.

**Comment by Czaki** [comment]
One important remark. Github actions do not allow to restart a single job. The previous provider allows this which simplifies debug. Also, non maintainer could restart the test on a previous platform.

**Comment by tlambert03** [comment]
thanks for pointing it out.  the fact that tests on the whole are finishing 3x faster kinda outweighs (for me at least) the fact that you just rerun the whole job.  There's [plenty of demand](https://github.community/t/ability-to-rerun-just-a-single-job-in-a-workflow/17234) for that feature too, so I suspect it will come soonish.  
Let's definitely keep an eye on how often we find a non-maintainer needing to restart a test without actually pushing a new commit (which, of course, is a way that anyone can restart a job)

**Pull request merged**
